drivers: stepper: tests: Test Generalization

jbehrensnx opened the pull request:
This PR includes the test generalization discussed in PR [#82843](https://github.com/zephyrproject-rtos/zephyr/pull/82843).

This PR has 3 objectives, first to offer a set of test suites for stepper drivers that can be used to help develop new drivers. The second objective is to offer a specification of the expected API behavior, something that is currently missing. And thirdly, to offer a tool to evaluate new drivers before approval.

While several stepper drivers have been developed, testing was often done only haphazardly, primarily using the stepper shell, such as PR [#79120](https://github.com/zephyrproject-rtos/zephyr/pull/79120). Testing via the shell is not only inconvenient and time consuming, it is also unreliable and non-repeatable, leading to many bugs and other issues to slip through. This set of test suites counteracts this by offering a set of repeatable tests that cover a wide range of expected behavior, making development easier and improving driver quality.

The stepper API currently lacks a specification for expected behavior. This not only makes driver development more difficult, as it is unclear how the new driver should behave in certain situations, it also leads to differences in behavior between drivers. This is in stark contrast to the goals of an API, where the user shouldn't care about the specific implementation and instead each api call should lead to the same result. These tests can alleviate this by defining expected behavior. Drivers that successfully pass them behave the same in the covered areas, improving their interchangeability from the user perspective.

Thirdly, does these test suites offer a way to evaluate new drivers before approval and thus ensure a certain level quality. Ideally, a new stepper should pass all applicable tests before it is approved.

## Test Suites

### stepper_api

This is a complete replacement of the existing stepper_api test suite. It checks general behavior of the driver and the results of incorrect input.

### stepper_const_velocity

This test suite checks that drivers that don't use acceleration and thus run at a constant velocity, move as expected. This includes that `move_to` and `move_by` finish in the expected time, and that all three movement commands result in the expected position after a given time.

## Current Test Results

All Tests were executed on a Nucleo F767ZI board with the `CONFIG_ZTEST_SHUFFLE` flag set, ensuring both random order and repetition. For the GPIO Stepper Controller, incorrect spinlock usage results in `test_micro_step_res_set_incorrect_value_detection` to crash the system. For this driver the test is skipped and considered to fail. The following table gives the test pass rate for both test suites and all tested drivers.

| Driver | stepper_api | stepper_const_velocity |
| ------ | ----------- | ---------------------- |
| GPIO Stepper Controller | 83.33% | 100% |
| DRV8424 Work Queue | 88.89% | 60% |
| DRV8424 Counter | 100% | 100% |
| TMC2209 Work Queue | 50% | 60% |
| TMC2209 Counter | 55.56% | 100%|

## Limitations

Like all tests, this set of test suites has its limitations. While the new `stepper_api` test suite should be applicable to all stepper driver, the `stepper_const_velocity`, as the name suggests, is only meant for drivers that move at a constant velocity. Drivers that use acceleration, such as tmc50xx or the acceleration step-dir implementation currently under development require their own test suite to evaluate movement.

Another limitation of this set of test suites is that they are limited to evaluating api return values. They can not check the physical behavior. In the case of a step-dir driver for example, they can not check, if the order is given for 100 steps, that the step signal actually send 100 pulses and that they occurred with the right frequency. Twisters support for these kinds of tests is currently limited. In addition, they require a far greater degree of customization for the target driver, going beyond the scope of this PR.

## Possible Future Work

- Fixing existing stepper drivers so that they pass all tests.
- Development of tests that evaluate actual outputs. Likely to be fairly board/driver specific, but some generalization for step-dir might be possible.
- Development of tests for drivers that use acceleration. Very high likelihood of being driver specific.
- Discussion (possibly in this PR) about acceptance criteria for new stepper drivers.

faxe1008 commented on tests/drivers/stepper/stepper_const_velocity/Kconfig:
```diff
@@ -6,4 +6,7 @@ config ENTROPY_GENERATOR
 	default y
 
+config SPEED_STEPS
+	int "Test speed in microsteps/s, also used as number of steps to make for certain tests"
```
a bit long and a grammar error, move this to a help text maybe?

faxe1008 commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -0,0 +1,24 @@
+config ENTROPY_GENERATOR
+	depends on ZTEST_SHUFFLE
+	bool "Shuffle the order of tests and suites"
+	default y
```
This looks like its reverse - shuffling would depend on the entropy generator no?

faxe1008 commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2024 Navimatix GmbH
```
Its 2025 :^)

jbehrensnx replied:
Most of it is reused from the drv8424 tests, which are from 2024. That said, it would be no problem to change 2024 to 2025,  the bigger question is just which copyrights to update.

faxe1008 commented on tests/drivers/stepper/stepper_api/boards/mimxrt1060_evk_mimxrt1062_qspi_B.overlay:
```diff
@@ -38,7 +32,7 @@
 
 / {
-	drv8424: drv8424 {
+	stepper_motor: stepper_motor{
```
idea: maybe make a `zephyr,chosen`  `stepper` node for it?

faxe1008 replied:
Add a:
```
zephyr,chosen {
      stepper = drv8424;
}
```

and use `DT_CHOSEN` within the test.

faxe1008 reviewed (commented):
First rough pass, thanks @jbehrensnx amazing effort!

jbehrensnx commented:
@jilaypandya  For some reason you were not included automatically

jilaypandya commented:
Thanks @jbehrensnx great work in unifying the tests :)

>  The stepper API currently lacks a specification for expected behavior. 

Should we not fix the stepper api specification before writing tests, i mean it's probably a change to the API documentation, or?

jbehrensnx commented:
> Thanks @jbehrensnx great work in unifying the tests :)
> 
> > The stepper API currently lacks a specification for expected behavior.
> 
> Should we not fix the stepper api specification before writing tests, i mean it's probably a change to the API documentation, or?

On some level you could say yes. The tests can however help us find cases where the current api specification is insufficient. That said, I would agree that we should complete the specification discussion before merging/approving this PR.

The other question is, should we discuss the api specification in this PR or in a RFC?

jbehrensnx commented:
- Fixed and added copyright information
- Temporarily removed qemu_x86_64 from testcases (overlays remain) as the stepper_api tests currently fail for this platform, thus failing some github checks
- Fixed wrong expected return value in one test (setting incorrect microstep resolution). This causes 1 test for tmc2209 and drv8424 to fail that previously passed and the inverse to happen for the fixed gpio driver

jilaypandya commented:
Please rebase :)

jbehrensnx commented:
Significant rework, incorporating the changes that have occurred since. Replaced test_config.yaml with tags. Usage of alias instead of chosen, as that is what the stepper_api test suite was using.

jilaypandya commented on tests/drivers/stepper/stepper_const_velocity/README.rst:
```diff
@@ -0,0 +1,79 @@
+suites is intended to be identical.
+
+
+Sample Output
```
This sample output is too redundant imo, please drop it. I think it's okay to have it in samples since it works as kind of an application, but in case of tests, they all have to pass.

jilaypandya commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -0,0 +1,23 @@
+	int "Test speed in microsteps/s"
+	default 50
+	help
+	  Besides being the test speed, this value is also used in several tests as the number of
```
Description says otherwise

jilaypandya commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -0,0 +1,23 @@
+	int "Microstep-resolution for testing of correct ms setting"
+	default 4
+
+config SPEED_STEPS
```
```suggestion
config USTEPS_PER_SECOND
```

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/mimxrt1060_evk_mimxrt1062_qspi_B/adi_tmc2209_workqueue.overlay:
```diff
@@ -0,0 +1,31 @@
+		step-gpios = <&arduino_header 10 0>; /* D4 */
+		en-gpios = <&arduino_header 11 0>; /* D5 */
+		msx-gpios = <&arduino_header 12 GPIO_ACTIVE_HIGH>,
+				<&arduino_header 13 GPIO_ACTIVE_HIGH>;
```
```suggestion
			<&arduino_header 13 GPIO_ACTIVE_HIGH>;
```

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/native_sim/adi_tmc2209_workqueue.overlay:
Can we have the moving of files in a separate commit?

jbehrensnx replied:
I can see what I can do, but as squashing everything into one commit was not the last thing I did, it might be difficult.

jilaypandya replied:
Right now lot of different things are squashed into one commit. Let's bring the changes step by step in.
In the first step probably just transfer the existing drv8424_api tests to stepper_api. That would also make reviewing easier and give more context about what is happening.

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/nucleo_f767zi/adi_tmc2209.overlay:
```diff
@@ -0,0 +1,34 @@
+};
+
+/ {
+	motor_1: motor_1 {
```
Let's just use the name of the controller as the node
```suggestion
	adi_tmc2209: adi_tmc2209 {
```

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/nucleo_f767zi/ti_drv8424.overlay:
```diff
@@ -38,21 +38,19 @@
 		status = "okay";
 		compatible = "ti,drv8424";
 
+
```
```suggestion
```

jilaypandya reviewed (commented):
While i do appreciate this extensive testing, for an api that is still experimental this would mean each time a change is introduced, quite a bit of refactoring will have to be undetaken

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/mimxrt1060_evk_mimxrt1062_qspi_B/adi_tmc2209_workqueue.overlay:
Is this actually tested on the hardware or is it just suggestive

jbehrensnx replied:
The drv8424, tmc2209 and a4979 drivers dont actually need hardware at this point. One can simply define and configure a node for them and then run the tests successfully. Which is what I did for the tmc and allegro.
I defined devicetree nodes for the drivers and tested them on the two boards. The wording is admittedly maybe a bit misleading, as I tested them on actual boards, just without an actual stepper controller.

jbehrensnx replied:
It is basically the same as the native_sim tests, just on a real board and not a simulated one.

jilaypandya reviewed (commented):
I think there are some overlays which you have not actually tested on the hardware, can we please not include them? I stick to the example of `nucleo_g071rb` since i test few drivers that i have with that board. 

Also this leaves room for further contributors who want to get their first PR badge :-)

```
Nucleo F767ZI
-------------

NOTE: For this test no real tmc2209 was used and thus the following pin configuration is
theoretical.
```

jilaypandya commented on tests/drivers/stepper/stepper_const_velocity/src/main.c:
Is this some pre-work for the upcoming ramp drivers?

jbehrensnx replied:
More or less, plus the tmc50xx driver already exists. As we have drivers with differing movement behavior, separating these tests seemed like a good idea.

jilaypandya replied:
so `set_micro_step_interval` is the only way of setting speed in stepper api. tmc50xx in its current form does not support this, the timing of the steps is decided internally by the IC. One can set the ramp parameters and the rest of the work is taken over by the IC.

so stepper api inherently does not have any functions which would suggest anything other than a fixed step interval ticking i.e. `set_micro_step_interval`.  Hence there is no need to diverge into two different test-suites as of now.

jbehrensnx replied:
Actually there is. `stepper_api` is intended to include tests that all stepper drivers need to be able to pass. This includes tmc50xx. The remaining functionality needs to be tested in other testsuites. This is where `stepper_const_velocity` comes into play. Another test suite concerning the specific behavior of tmc50xx should also be developed at some point, but that is beyond the scope of this PR.

jilaypandya commented on tests/drivers/stepper/stepper_const_velocity/testcase.yaml:
```diff
@@ -0,0 +1,124 @@
+    extra_args:
+      - platform:native_sim/native/64:DTC_OVERLAY_FILE="boards/native_sim/allegro_a4979.overlay"
+      - platform:nucleo_f767zi/stm32f767xx:DTC_OVERLAY_FILE="boards/nucleo_f767zi/allegro_a4979.overlay"
+      - platform:mimxrt1060_evk@B/mimxrt1062/qspi:DTC_OVERLAY_FILE="boards/mimxrt1060_evk_mimxrt1062_qspi_B/allegro_a4979.overlay"
```
please only add the overlays if they are actually tested on the hardware 

```
NOTE: For this test no real a4979 was used and thus the following pin configuration is
theoretical.
```

jbehrensnx commented:
I will drop the qemu and native sim platforms from the `const_velocity` testsuite, as they will always fail, and this causes issues with the automated tests.
For `stepper_api` I will remove the native_sim and qemu platforms from some tests for the same reason, but as in this case the failures are a result of the drivers themself and not timer issues in the simulation, the overlays and everything else will remain, so that they can be reenabled once the issues of the drivers are fixed.

jbehrensnx commented:
> While i do appreciate this extensive testing, for an api that is still experimental this would mean each time a change is introduced, quite a bit of refactoring will have to be undetaken

That is actually a very good reason to write all these tests, as they will help to ensure, that after these changes the api still behaves as expected. It is not quite Test Driven Development, but it borrows at least a few ideas from it.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +156,210 @@ ZTEST_F(stepper, test_stop)
+	int32_t steps = 100;
+	bool moving = true;
+
+	(void)stepper_set_microstep_interval(fixture->dev, fixture->test_interval);
```
That constant is intended to be used to test that a new microstep resolution can be successfully set. I forgot to add the test during the rework, which will be fixed. Managing the interval via a constant is however something I will do.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +156,210 @@ ZTEST_F(stepper, test_stop)
+	(void)stepper_move_by(fixture->dev, steps);
+	(void)stepper_enable(fixture->dev, false);
+	(void)stepper_is_moving(fixture->dev, &moving);
+	zassert_false(moving, "Driver should not be in state is_moving after being disabled");
```
Your recommendation is somewhat imprecise. There is a difference between the stepper being in state `is_moving` versus actually moving (i.e. position counter changing). The new error message could indicate the latter, which is however tested in another test.

jbehrensnx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +156,210 @@ ZTEST_F(stepper, test_stop)
+	(void)stepper_enable(fixture->dev, true);
+	(void)k_msleep(100);
+	(void)stepper_is_moving(fixture->dev, &moving);
+	zassert_false(moving, "Driver should not be in state is_moving after being reenabled");
```
Being reenabled is actually an important part of the test and thus should occur in the error message.

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +163,195 @@ ZTEST_F(stepper, test_stop)
+	zassert_false(moving, "Driver should not be in state is_moving after completion");
+}
+
+ZTEST_F(stepper_api, test_move_by_no_movement_when_disabled)
```
```suggestion
ZTEST_F(stepper_api, test_move_by_no_cancelled_when_disabled)
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +163,195 @@ ZTEST_F(stepper, test_stop)
+	(void)stepper_run(fixture->dev, STEPPER_DIRECTION_POSITIVE);
+	(void)stepper_is_moving(fixture->dev, &moving);
+	zassert_true(moving, "Driver should be in state is_moving");
+	(void)stepper_enable(fixture->dev, false);
```
```suggestion
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -166,3 +163,195 @@ ZTEST_F(stepper, test_stop)
+	(void)stepper_enable(fixture->dev, false);
+}
+
+ZTEST_F(stepper_api, test_run_no_movement_when_disabled)
```
```suggestion
ZTEST_F(stepper_api, test_run_cancelled_when_disabled)
```

jbehrensnx commented on tests/drivers/stepper/stepper_api/README.rst:
```diff
@@ -0,0 +1,31 @@
+#################
+
+Tests the general functionality and behaviour of stepper drivers to ensure conformity to the stepper
+api specification. All stepper drivers should pass this test suite. It should be noted, that this
```
The iterations in the api are exactly the reason why this test suite is important. It allows us to check both before and after the api change that the drivers behave as intended and that there are no unintended side effects. 

Constantly monitoring all stepper drivers that they conform to the test suite is however indeed impossible. Thus best effort and if possible occasional test runs have to suffice for harder to test drivers.

The line that all drivers should pass the test suite is intended for new drivers, so, admittedly, the wording might not be ideal. That said, the intention of the test suite is to check for conformity to the stepper api specification, and so a new driver that fails tests does not conform to the api specification and thus should not be accepted. With other words, while the wording admittedly did not make it clear, my intention was for this test suite to be an acceptance criteria for new drivers.

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/mimxrt1060_evk_mimxrt1062_qspi_B/zephyr_gpio_stepper.overlay:
What's the upside of adding these overlays for different boards?

jbehrensnx commented on tests/drivers/stepper/stepper_api/README.rst:
```diff
@@ -0,0 +1,31 @@
+#################
+
+Tests the general functionality and behaviour of stepper drivers to ensure conformity to the stepper
+api specification. All new stepper drivers should pass this test suite. It should be noted, that this
```
I am aware of this. That said, having a test suite that every new driver should pass (excluding exceptional circumstances with a good reason, I am not ruling these out) would improve the quality and consistency of new stepper drivers.

There are multiple possible reasons why other driver subsystems have not done this. One of them is finding a common core of functionality to set as a criteria (which can be seen even with stepper drivers, considering I split it in 2 test suites). The counter drivers for example have an extensive test suite, but it is missing tests for some functions while having tests not every driver can pass.
Another reason is that it might be very difficult to test these drivers in a meaningful way.
Then there the large number of existing drivers that might fail, which would result in many changes to these. In that regard, the low number of stepper drivers is a boon, as it makes this quite manageable.
And finally, people might not have want/been able to spend the required resources on this.

Another thing to consider is that with the growing commercial importance of Zephyr, testing will become more important and the stepper api could get a headstart on this.

Also, the goal of these tests is to test that a driver fulfills the api specification. If a new driver fails these tests, it does not fulfill the specification, so why would we want to accept it?

If we do it this way, there will be followup discussions about issues we find and about clarifications, but I think it is an important first step.

jilaypandya replied:
> And finally, people might not have want/been able to spend the required resources on this.

This is a very good reason to not increase the entry bar for new drivers.

jbehrensnx replied:
> > And finally, people might not have want/been able to spend the required resources on this.

> This is a very good reason to not increase the entry bar for new drivers.

This was in reference to spending the resources creating such a test suite.
Making the test suite mandatory doesn't really increase the entry bar for new drivers noticeably. In fact, one could argue that it lowers it. When developing a new driver you want to use the test suite during development anyway, as it is an easy way to check that your driver is doing the correct things. One will still need to use the shell on occasion, but the tests make the development process much easier. And with this background, succeeding the test suite becomes a normal part of developing a driver.  Thus not raising the entry bar.
In the end, we have something akin to Test Driven Development, but with the tests already existing.

jilaypandya replied:
> I am aware of this. That said, having a test suite that every new driver should pass (excluding exceptional circumstances with a good reason, I am not ruling these out)

What would be these exceptional circumstances?

I have mentioned this in one of the earlier comments as well, testing the purely gpio-based drivers is easy and as far as i can see can be very well covered by such test suites. But drivers like tmc5xxx or for that matter even tmc2209(only uart based) would need some sort of emulator. So what would we do in this case?

jbehrensnx replied:
> What would be these exceptional circumstances?

I'm not sure until it happens. As these tests are not perfect, there is always the possibility that there is some edge case or another reason that I did not consider In that case, these will need to be discussed. That is what I meant to say with this.

> I have mentioned this in one of the earlier comments as well, testing the purely gpio-based drivers is easy and as far as i can see can be very well covered by such test suites. But drivers like tmc5xxx or for that matter even tmc2209(only uart based) would need some sort of emulator. So what would we do in this case?

Those that can be tested via `native_sim/qemu` are, the rest can't be tested this way. That is a limitation of Zephyr, that some things can only be tested on real hardware. If however, someone develops a new driver, they have the means to test it. In that case we either have to trust their word or ask for the test log. Its one of the reasons, why only new tests should need to pass the test suite, as at that point there is somebody that can test the driver, later on that is however no longer the case.
The ideal solution in these cases would be for someone to develop an emulator, or at least stubbs for testing under `native_sim/qemu` but that is a lot of work and would, as you mentioned, raise the bar for entry, which is why I consider that nice to have but not required (and there is an argument to be made, that these emulators should be developed by the hardware developers, but that is a whole other discussion).

jbehrensnx committed:
drivers: stepper: Adds input data check for move_by

Adds a check for 0 steps to take for move_by in both
step_dir_stepper_common and the h-bridge driver.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

jbehrensnx committed:
drivers: stepper: tests: Transfers drv84xx api tests to stepper_api

Moves the tests of the drv84xx/api test suite into the stepper_api test
suite. To facilitate this and for cleanup purposes, several tests are
modified.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

jbehrensnx committed:
drivers: stepper: tests: Removes superfluous Tests

Removes the drv84xx/api test suite as well as a duplicate test in
stepper_api.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

dipakgmx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -122,87 +141,257 @@ ZTEST_F(stepper, test_actual_position)
 {
-	int32_t pos = 10u;
-	int ret;
+	(void)stepper_enable(fixture->dev);
```
I think, a better check of these calls is necessary, rather than void-ing out the returns. You can do something like:

`zassert_equal(stepper_enable(fixture->dev), 0, "Failed to enable device");`

jbehrensnx replied:
I don't think it is still necessary. Enable only charges the coils and has no impact on the stepper behavior otherwise. That said, a dedicated test for checking enable/disable would be an option. Checking it in every test case would be overkill imo.

jilaypandya replied:
One should always assert the return values of function calls in tests.
you can do something like this
`zassert_not_equal(stepper_enable(fixture->dev), -EIO, "Failed to enable device");`, coz if the enable returns `-ENOTSUP` or 0, the test-suite shall pass.

you can move enable in before respectively, so that you don't have to repeat it in each and every test.

@dipakgmx thanks for pointing this out, i will make sure i do this as well.

jilaypandya commented on tests/drivers/stepper/stepper_api/Kconfig:
```diff
@@ -11,3 +11,18 @@ config STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT
+	  As not all drivers support all microstep resolutions, so this needs to be adjusted
+	  on a per drvier basis.
+
+config STEPPER_TEST_ACCELERATION
```
```suggestion
```
We dont have acceleration support yet in stepper driver subsystem

jbehrensnx replied:
But we do have drivers that do use acceleration (i.e. tmc5xxx). Using this flag I skip certain tests that do not work with drivers that use acceleration. It replaces the `ret= -ENOSYS` test to filter them.

I originally used `STEPPER_TEST_CONST_VELOCITY` and checked for its existence to run said tests, but later thought that this might be the better solution. That said, I have no issue to switching back to `STEPPER_TEST_CONST_VELOCITY` if you think that that is better, but I definitly think that we should have a way to filter tests for this at this point in time.

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -122,87 +141,257 @@ ZTEST_F(stepper, test_actual_position)
+
+	(void)stepper_enable(fixture->dev);
+	(void)stepper_set_microstep_interval(fixture->dev, TEST_STEP_INTERVAL);
+	(void)stepper_set_event_callback(fixture->dev, fixture->callback, NULL);
```
> Also, before somebody asks, I removed the LOG_DBG in stepper_print_event_callback because it would cause native_sim to crash.

LOG_DBG in itself is not causing the native_sim to crash. Passing a NULL as user_data  here is causing the crash with stepper_print_event_callback
```suggestion
        (void)stepper_set_event_callback(fixture->dev, fixture->callback, (void *)fixture->dev);
```

dipakgmx replied:
> Also, before somebody asks, I removed the LOG_DBG in stepper_print_event_callback because it would cause native_sim to crash.

@jbehrensnx It's better to triage something locally before making bold statements like this. 
I was actually confused as to how `LOG_DBG` doesn't run in your case, where I keep using it when running tests and assumed this is some new changes with the testing framework.

jbehrensnx replied:
I meant the specific `LOG_DBG` statement, not `LOG_DBG` in general, sorry if that wasn't clear.
I probably should have done further experiments why the statement caused the error, but at that point I was just relieved to have found the source of the issue (I canceled a previous attempt at test generalization because I couldn't find the source of this issue). Commenting it out and doing further research would have been the better solution.

jilaypandya replied:
>  I removed the LOG_DBG in stepper_print_event_callback because it would cause native_sim to crash.

The crash should have happened on HW as well, because a NULL was getting passed around.

It's okay, sometimes things can be overlooked while developing, that's we get our work reviewed by each other and that's also the collaborative spirit of Open Source and Zephyr Project :). In general, I think we need to be conscious about the selection of words, for instance, this statement in the PR description, can come across as not so positive.
> While several stepper drivers have been developed, testing was often done only haphazardly, primarily using the stepper shell, such as PR 79120.

I highly appreciate the side effects which this PR have produced from introducing stepper_stop to improving API specification. Thanks for that :)

In general, there are Code Of Conduct guidelines in ZephyrProject https://github.com/zephyrproject-rtos/zephyr/blob/main/CODE_OF_CONDUCT.md#our-standards and we should abide by them.

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -55,9 +56,6 @@ static void stepper_print_event_callback(const struct device *dev, enum stepper_
 		break;
 	}
-
-	LOG_DBG("Event %d, %s called for %s, expected for %s\n", event, __func__,
```
This has to be something like, then LOG_DBG won't crash even if you pass NULL to user_data

```
LOG_DBG("Event %d, %s called for %s, expected for %s\n", event, __func__, dev_callback != NULL ? dev_callback->name : "",  dev->name);
```

jbehrensnx replied:
Added the parameter to the functions where it was NULL beforehand, but still used this version, because it is safer.

jilaypandya replied:
Thanks :)

jilaypandya commented:
> The `drivers.stepper.stepper_api.generic` issues can be seen in the failed tests as I would have to blacklist all boards the ci is run on, which is not an option. I am thinking about instead whitelisting (and adding an overlay for) a single board for this test scenario. This should fix the current issue with `.generic` and allow developers to use it with the `-K` twister flag.
> 
> If a physical board is still a no-go, then whitelisting a single emulation platform (maybe `qemu_x86_64/atom`, as it is already used for the h-bridge driver) with a conventional overlay file is also an option.

https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-2-inclusive-language

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -95,6 +108,19 @@ ZTEST_F(stepper, test_set_micro_step_res_invalid)
+ZTEST_F(stepper, test_set_micro_step_res_valid)
+{
+	int ret =
+		stepper_set_micro_step_res(fixture->dev, CONFIG_STEPPER_TEST_MICROSTEP_RESOLUTION);
```
This looks a bit strange. Most probably happened because of clang-format. 
Adjust the code as following.
```
	int ret; 
        ret = stepper_set_micro_step_res(fixture->dev, CONFIG_STEPPER_TEST_MICROSTEP_RESOLUTION);
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -112,97 +138,244 @@ ZTEST_F(stepper, test_set_micro_step_interval_invalid_zero)
+	zassert_equal(pos, 100, "Actual position should be %u but is %u", 100, pos);
 }
 
-ZTEST_F(stepper, test_target_position_w_fixed_step_interval)
```
Was this test similiar to a test in drv84xx test suite, and hence been deleted?

Please add the tests from drv84xx to stepper_api in one commit and then delete the duplicate test cases and do corresponding adjustments to stepper_api test suite in another commit. Thanks.

jbehrensnx replied:
Moved the removal to the drv84xx removal commit and renamed it to clarify that it removes superfluous tests.

Separating modifications to the tests into a separate  commit would be very difficult at this point, especially, as some changes are needed to run the tests. That said, I have adjusted the commit message to indicate that there also some modifications to tests.

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -112,97 +138,244 @@ ZTEST_F(stepper, test_set_micro_step_interval_invalid_zero)
 {
-	int32_t pos = 10u;
-	int ret;
+	(void)stepper_set_microstep_interval(fixture->dev, CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL);
```
One of the previous review comments about asserting the return values of stepper_api functions needs to be addressed. Thanks.

jbehrensnx replied:
Considering how often this function is called, I moved it into the `before` function and added a separate test for this. Because these tests are only for non-acceleration drivers, this should suffice.

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -112,97 +138,244 @@ ZTEST_F(stepper, test_set_micro_step_interval_invalid_zero)
-	bool is_moving;
+	(void)stepper_get_actual_position(fixture->dev, &steps);
+	int32_t min = target_position -
+		      ceil(((double)(CONFIG_STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT * 5)) / 100);
```
[DIV_ROUND_UP](https://docs.zephyrproject.org/apidoc/latest/group__sys-util.html#gae664e7492e37d324831caf2321ddda37) would work as well, or?

```suggestion
		      DIV_ROUND_UP(CONFIG_STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT * 5,  100);
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -112,97 +138,244 @@ ZTEST_F(stepper, test_set_micro_step_interval_invalid_zero)
-	}
+	(void)stepper_set_microstep_interval(fixture->dev, CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL);
+	(void)stepper_run(fixture->dev, STEPPER_DIRECTION_NEGATIVE);
+	k_msleep((-target_position) * CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL / NSEC_PER_MSEC);
```
```suggestion
	k_msleep(abs(target_position) * CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL / NSEC_PER_MSEC);
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -18,6 +19,10 @@ struct k_poll_signal stepper_signal;
 
+#define STEPPER_TIMEOUT(steps)                                                                     \
+	((steps * CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL / NSEC_PER_MSEC) *                        \
+	 (100 + CONFIG_STEPPER_TEST_TIMING_TIMEOUT_TOLERANCE_PCT) / 100)
```
This can result to zero, depending on CONFIG_STEPPER_TEST_MICROSTEP_INTERVAL is configured or? 
The best way to be add a `range 1000000 1000000000` in Kconfig for the parameter `STEPPER_TEST_MICROSTEP_INTERVAL`.
